aws_lambda_powertools/utilities/idempotency/idempotency.py

```diff
@@ -0,0 +1,101 @@
+"""
+Primary interface for idempotent Lambda functions utility
+"""
+import logging
+from typing import Any, Callable, Dict
+
+from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+from aws_lambda_powertools.utilities.idempotency.persistence.base import STATUS_CONSTANTS, BasePersistenceLayer
+
+from ..typing import LambdaContext
+from .exceptions import AlreadyInProgressError, ItemAlreadyExistsError, ItemNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+def default_error_callback():
+    raise
+
+
+@lambda_handler_decorator
+def idempotent(
+    handler: Callable[[Any, LambdaContext], Any],
+    event: Dict[str, Any],
+    context: LambdaContext,
+    persistence_store: BasePersistenceLayer,
+) -> Any:
+    """
+    Middleware to handle idempotency
+
+    Parameters
+    ----------
+    handler: Callable
+        Lambda's handler
+    event: Dict
+        Lambda's Event
+    context: Dict
+        Lambda's Context
+    persistence_store: BasePersistenceLayer
+        Instance of BasePersistenceLayer to store data
+
+    Examples
+    --------
+    **Processes Lambda's event in an idempotent manner**
+        >>> from aws_lambda_powertools.utilities.idempotency import idempotent, DynamoDBPersistenceLayer
+        >>>
+        >>> persistence_store = DynamoDBPersistenceLayer(event_key="body", table_name="idempotency_store")
+        >>>
+        >>> @idempotent(persistence_store=persistence_store)
+        >>> def handler(event, context):
+        >>>     return {"StatusCode": 200}
+    """
+
+    try:
+        # We call save_inprogress first as an optimization for the most common case where no idempotent record already
+        # exists. If it succeeds, there's no need to call get_record.
+        persistence_store.save_inprogress(event=event)
+    except ItemAlreadyExistsError:
+        try:
+            event_record = persistence_store.get_record(event)
+        except ItemNotFoundError:
+            return _call_lambda(handler=handler, persistence_store=persistence_store, event=event, context=context)
```

Comment:
@cakepietoast we might want to log a warning message here as this is an edge case where the record was deleted between the `save_inprogress` and `get_record` calls.

Reply:
@cakepietoast This might also raise an error?

Reply:
I'll add a log statement. This code path is unlikely to be hit often, but I can think of a couple of scenarios where it will occur (data layer expiring records or another concurrent execution raising an exception and deleting the record). In both of those cases, I think it makes sense proceed with executing the lambda handler.

Reply:
Shouldn't a call similar to save_inprogress (with extra error handling) happen here before the call to the handler? In this case another concurrent call can happen and execute the handler before this one is recorded duplicating the call.

Reply:
Yep, you're right - we need to retry the save_inprogress call here same as your other comment.